# HB-HMM
Extract the candidates of SNPs caused by deletion/LOH events using HMM model

## load sample data sets
```
load("sample_data/r.rda") # alternate allele
load("sample_data/cov.sc.rda") # total coverage
```

## load the script
```
source("HMM_code.R")
```

## set allele matrix
```
allele_matrix <- setAlleleMatrix(alter_sc = r, cov_sc = cov.sc, 
                                 het.deviance.threshold = 0.1, verbose = TRUE)
```
## plot allele-based information
```
allele_plot <- plotAlleleProfile(r.sub = allele_matrix$alter_less_sc, 
                                 n.sc.sub = allele_matrix$cov_sc, 
                                 l.sub = allele_matrix$alter_less_bulk, 
                                 n.bulk.sub = allele_matrix$cov_bulk, 
                                 snps = allele_matrix$snps)
```

## extract the candidates of continuous SNPs using HMM
```
allele_HMM <- calAlleleBoundaries(r.sub = allele_matrix$alter_less_sc, 
                                  n.sc.sub = allele_matrix$cov_sc)
```
## Update (2021/10/6):

## data preprocessing
```
alt.table <- read.table("sample_data/example.alt.matrix")
alt.table <- apply(alt.table, c(1,2), function(x) ifelse(x > 0 & x < 1, 0, x)) 
mean(alt.table > 0 & alt.table < 1)
dim(alt.table)

rownames(alt.table) <- gsub("::", ":", rownames(alt.table))

tot.table <- read.table("sample_data/example.tot.matrix") %>% as.matrix()
mean(tot.table > 0 & tot.table < 1)
dim(tot.table)

rownames(tot.table) <- gsub("::", ":", rownames(tot.table))
```

## create a object for allele-based method
```
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
test <- CreateInfercnvAlleleObject(allele_counts_matrix = alt.table,
                                   coverage_counts_matrix = tot.table,
                                   ref_annotation_file = txdb)
```

## set allele matrix
```
test <- setAlleleMatrix(test, het.deviance.threshold = 0.1)
```

## extract the candidates of continuous SNPs using HMM
```
test <- calAlleleBoundaries(test)
```
## HMM output
```
GRanges object with 24 ranges and 0 metadata columns:
       seqnames              ranges strand
          <Rle>           <IRanges>  <Rle>
   [1]    chr10     812082-84256410      *
   [2]    chr11 112045923-113240224      *
   [3]    chr11 122713154-124923820      *
   [4]    chr11     703595-72110190      *
   [5]    chr14   64442127-99909370      *
   ...      ...                 ...    ...
  [20]     chr5  10239166-109383450      *
  [21]     chr5     766098-80142489      *
  [22]     chr7 100038344-104419676      *
  [23]     chr7   1609484-159122104      *
  [24]     chrX    536656-155259147      *
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths
```

## Update (2021/10/16):
```
source("script/infercnv_allele.R")
file_name <- c("MGH36","MGH53", "MGH54")

res = list()
for(i in file_name){
  
  if(i == "MGH36"){
    ref_group_names = "Microglia/Macrophage" 
  }
  if(i %in% c("MGH53", "MGH54")){
    ref_group_names = c("Microglia/Macrophage","Oligodendrocytes (non-malignant)") 
  }
  
  test <- CreateInfercnvAlleleObject(allele_counts_matrix = paste0("processed_data/",i,".alt.dense.matrix"),
                                     coverage_counts_matrix = paste0("processed_data/",i,".tot.dense.matrix"),
                                     ref_annotation_file = "gencode_v19_gene_pos.txt",
                                     sample_annotation_file = paste0("processed_data/",i,".cell_types"),
                                     ref_group_names = ref_group_names,
                                     rowname_by = "::")
  test <- setAlleleMatrix(test, het.deviance.threshold = 0.1)
  test <- calAlleleBoundaries(test, distance_method = "Filter_threshold")
  plot_allele(test, name_to_plot = paste0(i,"_HMM_oct_seventeen_Filter_threshold.pdf"))
  test <- calAlleleBoundaries(test, distance_method = "Remove_NA")
  plot_allele(test, name_to_plot = paste0(i,"_HMM_oct_seventeen_Remove_NA.pdf"))
  test <- calAlleleBoundaries(test, distance_method = "HB")
  plot_allele(test, name_to_plot = paste0(i,"_HMM_oct_seventeen_HB.pdf"))
  
  res[[i]] <- test
}

file_name_2 <- c("MGH97","MGH60", "MGH93")
res_2 = list()
for(i in file_name_2){
  
  test <- CreateInfercnvAlleleObject(allele_counts_matrix = paste0("processed_data/",i,".alt.dense.matrix"),
                                     coverage_counts_matrix = paste0("processed_data/",i,".tot.dense.matrix"),
                                     ref_annotation_file = "gencode_v19_gene_pos.txt",
                                     sample_annotation_file = paste0("processed_data/",i,".cell_types"),
                                     ref_group_names = NULL,
                                     rowname_by = "::")
  test <- setAlleleMatrix(test, het.deviance.threshold = 0.1)
  test <- calAlleleBoundaries(test, distance_method = "Filter_threshold")
  plot_allele(test, name_to_plot = paste0(i,"_HMM_oct_seventeen_Filter_threshold.pdf"))
  test <- calAlleleBoundaries(test, distance_method = "Remove_NA")
  plot_allele(test, name_to_plot = paste0(i,"_HMM_oct_seventeen_Remove_NA.pdf"))
  test <- calAlleleBoundaries(test, distance_method = "HB")
  plot_allele(test, name_to_plot = paste0(i,"_HMM_oct_seventeen_HB.pdf"))
  
  res_2[[i]] <- test
}
```
## Running logs for file_name_2 ("MGH97","MGH60", "MGH93")
```
INFO [2021-10-17 22:31:52] Parsing matrix: processed_data/MGH97.alt.dense.matrix ...
INFO [2021-10-17 22:32:08] Parsing matrix: processed_data/MGH97.tot.dense.matrix ...
INFO [2021-10-17 22:32:22] Creating in-silico bulk ...
INFO [2021-10-17 22:32:22] Parsing gene annotation file: gencode_v19_gene_pos.txt ...
INFO [2021-10-17 22:32:23] Parsing cell annotation file: processed_data/MGH97.cell_types ...
INFO [2021-10-17 22:32:26] Validating infercnv_allele obejct ...
INFO [2021-10-17 22:32:27] 38803 heterozygous SNPs identified ...
INFO [2021-10-17 22:32:28] Setting composite lesser allele count ...
INFO [2021-10-17 22:32:30] Done setting initial allele matrices!
INFO [2021-10-17 22:32:30] Starting cluster cells from population ...
INFO [2021-10-17 22:32:30] Using Filter_threshold to process matrix ...
INFO [2021-10-17 22:32:35] Starting calculate distance ...
INFO [2021-10-17 22:32:47] Starting calculate Hierarchical Clustering ...
INFO [2021-10-17 22:32:47] Starting iterative HMM ...
INFO [2021-10-17 22:33:05] Done extracting HMM regions ...
INFO [2021-10-17 22:33:07] Building snp plot ...
INFO [2021-10-17 22:33:09] Number het snps used: 8808 ...
INFO [2021-10-17 22:33:09] Setting alt allele fraction to the tumor cell-population minor allele ...
INFO [2021-10-17 22:33:12] Melting matrix ...
INFO [2021-10-17 22:33:25] Generating outputs ...
INFO [2021-10-17 22:33:25] Making malignant plot ...
INFO [2021-10-17 22:33:25] Making allele freq plot ...
INFO [2021-10-17 22:33:30] Smoothing sample means for trend lines ...
INFO [2021-10-17 22:33:32] Making HMM prediction plot ...
INFO [2021-10-17 22:33:55] Done!
INFO [2021-10-17 22:33:55] Starting cluster cells from population ...
INFO [2021-10-17 22:33:55] Using Remove_NA to process matrix ...
INFO [2021-10-17 22:33:59] Starting calculate distance ...
INFO [2021-10-17 22:34:11] Starting calculate Hierarchical Clustering ...
INFO [2021-10-17 22:34:11] Starting iterative HMM ...
INFO [2021-10-17 22:34:31] Done extracting HMM regions ...
INFO [2021-10-17 22:34:33] Building snp plot ...
INFO [2021-10-17 22:34:35] Number het snps used: 8808 ...
INFO [2021-10-17 22:34:35] Setting alt allele fraction to the tumor cell-population minor allele ...
INFO [2021-10-17 22:34:38] Melting matrix ...
INFO [2021-10-17 22:34:49] Generating outputs ...
INFO [2021-10-17 22:34:49] Making malignant plot ...
INFO [2021-10-17 22:34:50] Making allele freq plot ...
INFO [2021-10-17 22:34:55] Smoothing sample means for trend lines ...
INFO [2021-10-17 22:34:58] Making HMM prediction plot ...
INFO [2021-10-17 22:35:14] Done!
INFO [2021-10-17 22:35:14] Starting cluster cells from population ...
INFO [2021-10-17 22:35:14] Using HB to process matrix ...
INFO [2021-10-17 22:35:17] Starting calculate distance ...
INFO [2021-10-17 22:37:08] Starting calculate Hierarchical Clustering ...
INFO [2021-10-17 22:37:08] Starting iterative HMM ...
INFO [2021-10-17 22:37:26] Done extracting HMM regions ...
INFO [2021-10-17 22:37:28] Building snp plot ...
INFO [2021-10-17 22:37:31] Number het snps used: 8808 ...
INFO [2021-10-17 22:37:31] Setting alt allele fraction to the tumor cell-population minor allele ...
INFO [2021-10-17 22:37:34] Melting matrix ...
INFO [2021-10-17 22:37:46] Generating outputs ...
INFO [2021-10-17 22:37:46] Making malignant plot ...
INFO [2021-10-17 22:37:46] Making allele freq plot ...
INFO [2021-10-17 22:37:51] Smoothing sample means for trend lines ...
INFO [2021-10-17 22:37:54] Making HMM prediction plot ...
INFO [2021-10-17 22:38:10] Done!
INFO [2021-10-17 22:38:10] Parsing matrix: processed_data/MGH60.alt.dense.matrix ...
INFO [2021-10-17 22:38:21] Parsing matrix: processed_data/MGH60.tot.dense.matrix ...
INFO [2021-10-17 22:38:31] Creating in-silico bulk ...
INFO [2021-10-17 22:38:32] Parsing gene annotation file: gencode_v19_gene_pos.txt ...
INFO [2021-10-17 22:38:32] Parsing cell annotation file: processed_data/MGH60.cell_types ...
INFO [2021-10-17 22:38:35] Validating infercnv_allele obejct ...
INFO [2021-10-17 22:38:37] 55835 heterozygous SNPs identified ...
INFO [2021-10-17 22:38:38] Setting composite lesser allele count ...
INFO [2021-10-17 22:38:38] Done setting initial allele matrices!
INFO [2021-10-17 22:38:38] Starting cluster cells from population ...
INFO [2021-10-17 22:38:39] Using Filter_threshold to process matrix ...
INFO [2021-10-17 22:38:44] Starting calculate distance ...
INFO [2021-10-17 22:38:56] Starting calculate Hierarchical Clustering ...
INFO [2021-10-17 22:38:56] Starting iterative HMM ...
INFO [2021-10-17 22:39:22] Done extracting HMM regions ...
INFO [2021-10-17 22:39:24] Building snp plot ...
INFO [2021-10-17 22:39:26] Number het snps used: 12710 ...
INFO [2021-10-17 22:39:26] Setting alt allele fraction to the tumor cell-population minor allele ...
INFO [2021-10-17 22:39:29] Melting matrix ...
INFO [2021-10-17 22:39:43] Generating outputs ...
INFO [2021-10-17 22:39:43] Making malignant plot ...
INFO [2021-10-17 22:39:43] Making allele freq plot ...
INFO [2021-10-17 22:39:49] Smoothing sample means for trend lines ...
INFO [2021-10-17 22:39:52] Making HMM prediction plot ...
INFO [2021-10-17 22:40:10] Done!
INFO [2021-10-17 22:40:10] Starting cluster cells from population ...
INFO [2021-10-17 22:40:11] Using Remove_NA to process matrix ...
INFO [2021-10-17 22:40:15] Starting calculate distance ...
INFO [2021-10-17 22:40:26] Starting calculate Hierarchical Clustering ...
INFO [2021-10-17 22:40:26] Starting iterative HMM ...
INFO [2021-10-17 22:40:53] Done extracting HMM regions ...
INFO [2021-10-17 22:40:55] Building snp plot ...
INFO [2021-10-17 22:40:56] Number het snps used: 12710 ...
INFO [2021-10-17 22:40:56] Setting alt allele fraction to the tumor cell-population minor allele ...
INFO [2021-10-17 22:41:00] Melting matrix ...
INFO [2021-10-17 22:41:11] Generating outputs ...
INFO [2021-10-17 22:41:11] Making malignant plot ...
INFO [2021-10-17 22:41:11] Making allele freq plot ...
INFO [2021-10-17 22:41:18] Smoothing sample means for trend lines ...
INFO [2021-10-17 22:41:21] Making HMM prediction plot ...
INFO [2021-10-17 22:41:39] Done!
INFO [2021-10-17 22:41:39] Starting cluster cells from population ...
INFO [2021-10-17 22:41:40] Using HB to process matrix ...
INFO [2021-10-17 22:41:43] Starting calculate distance ...
INFO [2021-10-17 22:43:08] Starting calculate Hierarchical Clustering ...
INFO [2021-10-17 22:43:08] Starting iterative HMM ...
INFO [2021-10-17 22:43:34] Done extracting HMM regions ...
INFO [2021-10-17 22:43:38] Building snp plot ...
INFO [2021-10-17 22:43:39] Number het snps used: 12710 ...
INFO [2021-10-17 22:43:39] Setting alt allele fraction to the tumor cell-population minor allele ...
INFO [2021-10-17 22:43:43] Melting matrix ...
INFO [2021-10-17 22:43:57] Generating outputs ...
INFO [2021-10-17 22:43:57] Making malignant plot ...
INFO [2021-10-17 22:43:57] Making allele freq plot ...
INFO [2021-10-17 22:44:06] Smoothing sample means for trend lines ...
INFO [2021-10-17 22:44:10] Making HMM prediction plot ...
INFO [2021-10-17 22:44:29] Done!
INFO [2021-10-17 22:44:29] Parsing matrix: processed_data/MGH93.alt.dense.matrix ...
INFO [2021-10-17 22:44:39] Parsing matrix: processed_data/MGH93.tot.dense.matrix ...
INFO [2021-10-17 22:44:52] Creating in-silico bulk ...
INFO [2021-10-17 22:44:53] Parsing gene annotation file: gencode_v19_gene_pos.txt ...
INFO [2021-10-17 22:44:53] Parsing cell annotation file: processed_data/MGH93.cell_types ...
INFO [2021-10-17 22:44:56] Validating infercnv_allele obejct ...
INFO [2021-10-17 22:44:58] 58831 heterozygous SNPs identified ...
INFO [2021-10-17 22:44:58] Setting composite lesser allele count ...
INFO [2021-10-17 22:44:59] Done setting initial allele matrices!
INFO [2021-10-17 22:44:59] Starting cluster cells from population ...
INFO [2021-10-17 22:44:59] Using Filter_threshold to process matrix ...
INFO [2021-10-17 22:45:05] Starting calculate distance ...
INFO [2021-10-17 22:45:18] Starting calculate Hierarchical Clustering ...
INFO [2021-10-17 22:45:18] Starting iterative HMM ...
INFO [2021-10-17 22:45:45] Done extracting HMM regions ...
INFO [2021-10-17 22:45:47] Building snp plot ...
INFO [2021-10-17 22:45:49] Number het snps used: 13862 ...
INFO [2021-10-17 22:45:49] Setting alt allele fraction to the tumor cell-population minor allele ...
INFO [2021-10-17 22:45:53] Melting matrix ...
INFO [2021-10-17 22:46:07] Generating outputs ...
INFO [2021-10-17 22:46:07] Making malignant plot ...
INFO [2021-10-17 22:46:07] Making allele freq plot ...
INFO [2021-10-17 22:46:15] Smoothing sample means for trend lines ...
INFO [2021-10-17 22:46:19] Making HMM prediction plot ...
INFO [2021-10-17 22:46:39] Done!
INFO [2021-10-17 22:46:39] Starting cluster cells from population ...
INFO [2021-10-17 22:46:40] Using Remove_NA to process matrix ...
INFO [2021-10-17 22:46:43] Starting calculate distance ...
INFO [2021-10-17 22:46:56] Starting calculate Hierarchical Clustering ...
INFO [2021-10-17 22:46:56] Starting iterative HMM ...
INFO [2021-10-17 22:47:25] Done extracting HMM regions ...
INFO [2021-10-17 22:47:28] Building snp plot ...
INFO [2021-10-17 22:47:29] Number het snps used: 13862 ...
INFO [2021-10-17 22:47:29] Setting alt allele fraction to the tumor cell-population minor allele ...
INFO [2021-10-17 22:47:33] Melting matrix ...
INFO [2021-10-17 22:47:46] Generating outputs ...
INFO [2021-10-17 22:47:46] Making malignant plot ...
INFO [2021-10-17 22:47:47] Making allele freq plot ...
INFO [2021-10-17 22:47:56] Smoothing sample means for trend lines ...
INFO [2021-10-17 22:48:00] Making HMM prediction plot ...
INFO [2021-10-17 22:48:21] Done!
INFO [2021-10-17 22:48:21] Starting cluster cells from population ...
INFO [2021-10-17 22:48:21] Using HB to process matrix ...
INFO [2021-10-17 22:48:27] Starting calculate distance ...
INFO [2021-10-17 22:49:57] Starting calculate Hierarchical Clustering ...
INFO [2021-10-17 22:49:57] Starting iterative HMM ...
INFO [2021-10-17 22:50:25] Done extracting HMM regions ...
INFO [2021-10-17 22:50:27] Building snp plot ...
INFO [2021-10-17 22:50:29] Number het snps used: 13862 ...
INFO [2021-10-17 22:50:29] Setting alt allele fraction to the tumor cell-population minor allele ...
INFO [2021-10-17 22:50:33] Melting matrix ...
INFO [2021-10-17 22:50:46] Generating outputs ...
INFO [2021-10-17 22:50:46] Making malignant plot ...
INFO [2021-10-17 22:50:47] Making allele freq plot ...
INFO [2021-10-17 22:50:54] Smoothing sample means for trend lines ...
INFO [2021-10-17 22:50:57] Making HMM prediction plot ...
INFO [2021-10-17 22:51:17] Done!
```
![alt text](sample_fig/MGH60_HMM_oct_seventeen_Filter_threshold.jpg "test1")
